crates/bevy_math/src/smooth.rs

```diff
@@ -0,0 +1,118 @@
+                // let exp = 1.0 / (1.0 + x * (1.0 + x * (0.48 + 0.235 * x))); // TODO: profile me, both in debug & release
+
+                let change = from - to;
+                let temp = (velocity + omega * change) * delta_time;
+
+                (
+                    to + (change + temp) * exp,      // position
+                    (velocity - omega * temp) * exp, // velocity
+                )
+            }
+        }
+    };
+}
+
+impl_smooth_damp! {f32, f32}
+impl_smooth_damp! {f64, f64}
+impl_smooth_damp! {Vec2, f32}
+impl_smooth_damp! {Vec3, f32}
+
+/// Smooths value to a goal using a damped spring limited by a maximum speed.
+pub trait SmoothDampMax {
+    /// Smooths value to a goal using a damped spring limited by a maximum speed.
+    ///
+    /// `smooth_time` is the expected time to reach the target when at maximum velocity.
+    ///
+    /// Returns smoothed value and new velocity.
+    fn smooth_damp_max(
+        from: Self,
+        to: Self,
+        velocity: Self,
+        max_speed: f32,
+        smooth_time: f32,
+        delta_time: f32,
+    ) -> (Self, Self)
+    where
+        Self: Sized;
+}
+
+macro_rules! impl_smooth_damp_max {
+    ($t:ty, $f:ty, $clamp:expr) => {
+        impl SmoothDampMax for $t {
+            fn smooth_damp_max(
+                from: $t,
+                to: $t,
+                velocity: $t,
+                max_speed: f32,
+                smooth_time: f32,
+                delta_time: f32,
+            ) -> ($t, $t) {
+                let max_speed = <$f>::max(max_speed as $f, 0.0001); // ensure max speed is positive and non-zero
+                let smooth_time = <$f>::max(smooth_time as $f, 0.0001); // ensure smooth_time is positive and non-zero
+                let delta_time = delta_time as $f;
+
+                // from game programming gems 4, chapter 1.10
+                let omega = 2.0 / smooth_time;
+                let x = omega * delta_time;
+
+                let exp = 1.0 / (1.0 + x + 0.48 * x * x + 0.235 * x * x * x);
+                // let exp = 1.0 / (1.0 + x * (1.0 + x * (0.48 + 0.235 * x))); // TODO: profile me, both in debug & release
+
+                let max = max_speed * smooth_time;
+                let change = from - to;
+                let change = $clamp(change, max);
+                let to = from - change;
+
+                let temp = (velocity + omega * change) * delta_time;
+
+                (
+                    to + (change + temp) * exp,      // position
+                    (velocity - omega * temp) * exp, // velocity
+                )
+            }
+        }
+    };
+}
+
+impl_smooth_damp_max! {f32, f32, |change, max:f32| { f32::clamp(change, -max, max) }}
+impl_smooth_damp_max! {f64, f64, |change, max:f64| { f64::clamp(change, -max, max) }}
+impl_smooth_damp_max! {Vec2, f32, |change:Vec2, max| { change.clamp_length_max(max) }}
+impl_smooth_damp_max! {Vec3, f32, |change:Vec3, max| { change.clamp_length_max(max) }}
```

Comment:
You should probably write some test cases in a test module `#[cfg(test)]`.
While I believe your implementation is correct, we should really start having tests for new feature we add to bevy.

Reply:
Yup. Usually math functions are great candidates for tests. However, I've been scratching my head since the beginning and I have no idea for a good test for this. It's usually used iteratively... Any suggestions?